File: 01-arrays_&_hashing/00347-Top_K_Frequent_Elements/00347-top_k_frequent_elements.go
// 347: Top K Frequent Elements
// https://leetcode.com/problems/top-k-frequent-elements/

package main

import (
	"fmt"
	"sort"
)

// SOLUTION
func topKFrequent(nums []int, k int) []int {
	tmap := make(map[int]int)
	var buckets []int
	
    for _, v := range nums {
        tmap[v]++
    }

    for k := range tmap {
        buckets = append(buckets, k)
    }
    
    sort.Slice(buckets, func (i, j int) bool {
        return tmap[buckets[i]] > tmap[buckets[j]]
    })
    
    return buckets[:k]
}


func main() {
	// INPUT
	nums := []int{1,1,1,2,2,3}
	k := 2

	// OUTPUT
	result := topKFrequent(nums, k)
	fmt.Println(result)
}

File: 01-arrays_&_hashing/00347-Top_K_Frequent_Elements/00347-top_k_frequent_elements.py
# 347: Top K Frequent Elements
# https://leetcode.com/problems/top-k-frequent-elements/

class Solution:
    # SOLUTION
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        count = {}
        buckets = [[] for _ in range(len(nums) + 1)]

        for n in nums:
            count[n] = 1 + count.get(n, 0)
        for n, c in count.items():
            buckets[c].append(n)

        result = []
        for i in range(len(buckets)-1, 0, -1):
            for n in buckets[i]:
                result.append(n)
            if len(result) == k: break

        return result


if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums = [1,1,1,2,2,3]
    k = 2

    # OUTPUT
    result = o.topKFrequent(nums, k)
    print(result)
